### PR TITLE
Support updated reputation contract

### PR DESCRIPTION
**Pull request by alexvanin**
Related to https://github.com/nspcc-dev/neofs-contract/pull/65

- [x] Support reputation contract in morph/client
- [x] Add client wrapper
- [x] Support reputation put event 
- [x] Add reputation processor to inner ring node
- [x] Add comments
- [x] Use structure to represent reputation value instead of `[]byte`.

### CODE REVIEW

File: pkg/innerring/processors/reputation/process_put.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	"encoding/hex"
+
+	"github.com/nspcc-dev/neofs-api-go/pkg/reputation"
+	"github.com/nspcc-dev/neofs-node/pkg/morph/client/reputation/wrapper"
+	"go.uber.org/zap"
+)
+
+func (rp *Processor) processPut(epoch uint64, id reputation.PeerID, value reputation.GlobalTrust) {
+	if !rp.alphabetState.IsAlphabet() {
+		rp.log.Info("non alphabet mode, ignore new peer notification")
```
Comment (cthulhu-rider) [inline pkg/innerring/processors/reputation/process_put.go:13]: I guess incorrect message

File: pkg/innerring/processors/reputation/process_put.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		return
+	}
+
+	// check if epoch is valid
+	currentEpoch := rp.epochState.EpochCounter()
+	if epoch >= rp.epochState.EpochCounter() {
```
Comment (cthulhu-rider) [inline pkg/innerring/processors/reputation/process_put.go:19]:
```suggestion
	if epoch >= currentEpoch {
```

File: pkg/morph/client/reputation/get.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+func (g *GetArgs) SetPeerID(v []byte) {
+	g.peerID = v
+}
+
+// SetID sets id of expected reputation value in reputation contract.
+func (g *GetByIDArgs) SetID(v []byte) {
+	g.id = v
+}
+
+// Reputations returns marshaled slice of reputation values.
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/get.go:42]:
```suggestion
// Reputations returns slice of marshaled reputation values.
```

File: pkg/morph/client/reputation/put.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+func (p *PutArgs) SetPeerID(v []byte) {
+	p.peerID = v
+}
+
+// SetValue sets marshaled reputation value.
+func (p *PutArgs) SetValue(v []byte) {
+	p.value = v
+}
+
+// Put invokes direct call of "put reputation value" method of audit contract.
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/put.go:29]:
```suggestion
// Put invokes direct call of "put reputation value" method of reputation contract.
```

File: pkg/morph/client/reputation/put.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	return errors.Wrapf(c.client.Invoke(
+		c.putMethod,
+		int64(args.epoch),
+		args.peerID,
+		args.value,
+	), "could not invoke method (%s)", c.putMethod)
+}
+
+// PutViaNotary invokes notary call of "put reputation value" method of
+// audit contract.
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/put.go:40]:
```suggestion
// reputation contract.
```

File: pkg/morph/client/reputation/put.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+func (c *Client) PutViaNotary(args PutArgs) error {
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/put.go:41]: Cosmetic: add boolean `notary` to `PutArgs` and leave only method `Put` (the same in `wrapper` pkg).

File: pkg/morph/client/reputation/wrapper/get.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+		peerID reputation.PeerID
+	}
+
+	// GetByIDArgs groups the arguments of "get reputation value by
+	// reputation id" test invocation.
+	GetByIDArgs struct {
+		id ReputationID
+	}
+
+	// ListByEpochResult groups the result of "get reputation value" and
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/wrapper/get.go:22]:
```suggestion
	// GetResult groups the result of "get reputation value" and
```

File: pkg/morph/client/reputation/wrapper/get.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+// Reputations returns slice of reputation values.
+func (g GetResult) Reputations() []reputation.GlobalTrust {
+	return g.reputations
+}
+
+// Get invokes the call of "get reputation value" method of reputation contract.
+func (w *ClientWrapper) Get(v GetArgs) (*GetResult, error) {
+	args := reputationClient.GetArgs{}
+	args.SetEpoch(v.epoch)
+	args.SetPeerID(v.peerID.ToV2().GetValue())
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/wrapper/get.go:53]: Why don't we use `v.peerID.Marshal()` (same in `wrapper.preparePutArgs`)?

File: pkg/morph/client/reputation/wrapper/put.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+type (
+	// PutArgs groups the arguments of "put reputation value" invocation call.
+	PutArgs struct {
+		epoch  uint64
+		peerID reputation.PeerID
+		value  reputation.GlobalTrust
+	}
+)
+
+// SetPeerID sets epoch of reputation value.
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/wrapper/put.go:18]:
```suggestion
// SetEpoch sets epoch of reputation value.
```

File: pkg/morph/client/reputation/wrapper/put.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+func (p *PutArgs) SetPeerID(v reputation.PeerID) {
+	p.peerID = v
+}
+
+// SetValue sets reputation value.
+func (p *PutArgs) SetValue(v reputation.GlobalTrust) {
+	p.value = v
+}
+
+// Put invokes direct call of "put reputation value" method of audit contract.
```
Comment (cthulhu-rider) [inline pkg/morph/client/reputation/wrapper/put.go:33]:
```suggestion
// Put invokes direct call of "put reputation value" method of reputation contract.
```

File: pkg/morph/client/reputation/put.go (hunk 6 above)
Comment (alexvanin) [reply pkg/morph/client/reputation/put.go:41]: I would like to keep two separate functions. Notary call is an add-on for client and it should not affect functions that can be used without notary, I think.

File: pkg/morph/client/reputation/wrapper/get.go (hunk 8 above)
Comment (alexvanin) [reply pkg/morph/client/reputation/wrapper/get.go:53]: PeerID value used as a part of the key for storage in reputation contract. Storage key has it's own size limits, so I decided to keep exact value right there without extra meta information of protobuf encoding. This should encourage developers to not change `reputation.PeerID` definition in API.

Comment (cthulhu-rider) [reply pkg/morph/client/reputation/wrapper/get.go:53]: But why contract's storage key imposes requirements to `PeerID` size? It becomes pretty version-dependent then, same as this part of the code.

Comment (alexvanin) [reply pkg/morph/client/reputation/wrapper/get.go:53]: > But why contract's storage key imposes requirements to PeerID size? 

It is NEO limitation of max storage key size (64 byte), so we can't do anything about that. 

Another thing is that bytes of marshaled `PeerID` structure is not PeerID. The same as bytes of marshaled structure of ContainerID is not ContainerID, so it seemed legit to use value in this case.

Comment (cthulhu-rider) [reply pkg/morph/client/reputation/wrapper/get.go:53]: I got it, in the background I will think about how to simultaneously adjust to the restrictions of the contract and avoid using `ToV2()` (apparently some way of getting a binary ID of a fixed size).

**Review by cthulhu-rider — APPROVED** [review]

**Pull request merged**